src/github.com/matrix-org/dendrite/syncserver/storage/output_room_events_table.go

```diff
@@ -47,8 +47,10 @@ const selectRecentEventsSQL = "" +
 const selectMaxIDSQL = "" +
 	"SELECT MAX(id) FROM output_room_events"
 
+// In order for us to apply the state updates correctly, rows need to be ordered in the order they were received (id).
 const selectStateInRangeSQL = "" +
-	"SELECT event_json, add_state_ids, remove_state_ids FROM output_room_events WHERE id > $1 AND id < $2 AND add_state_ids IS NOT NULL OR remove_state_ids IS NOT NULL"
+	"SELECT event_json, add_state_ids, remove_state_ids FROM output_room_events WHERE id > $1 AND id < $2 AND " +
+	"add_state_ids IS NOT NULL OR remove_state_ids IS NOT NULL ORDER BY id ASC"
```

Comment:
Also do you remember what the precedence rules for `AND` and `OR` are in sql? I don't so I kind of prefer to use parenthesises to save myself the trouble of having to look them up.